docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,844 @@
+contract, it is possible for the operator to unpause providers; the operator
+does not allow to unpause providers if there are still providers abiding to
+the old contract.
+
+Please note that we are planning to embed this sequence (pause - upgrade -
+unpause) as a part of `clusterctl upgrade apply` command when there is a
+contract change.
+
+
+##### Changing a provider
+
+On top of changing a provider version (upgrades), the operator supports also
+changing other provider fields, most notably controller flags and variables.
+
+The operation internally works like upgrades: The current instance of the
+provider is deleted, while preserving CRDs, namespaced and user objects A new
+instance of the provider is installed with the new set of flags/variables.
+
+Please note that clusterctl currently does not support this operation.
+
+See Example 1 in [Example API Usage](#example-api-usage)
+
+##### Deleting a provider
+
+In order to delete a provider you have to delete the corresponding provider
+object.
+
+Deletion of the provider will be blocked if any workload cluster using the
+provider still exists.
+
+Additionally, deletion of a core provider should be blocked if there are still
+other providers in the management cluster.
+
+#### Operator Lifecycle Management
+
+##### Operator Installation
+
+- `clusterctl init` will install the operator and its corresponding CRDs as a
+  pre-requisite if the operator doesn’t already exist. Please note that this
+  command will consider image overrides defined in the local clusterctl config
+  file.
+- If the admin does not want to use clusterctl to install the operator, it is
+  possible to kubectl apply the operator yaml that will be published in the
+  cluster-api release artifacts.
+- There can be only one operator installed per management cluster, across all
+  namespaces.
+
+##### Operator Upgrade
+
+- The admin can use `clusterctl upgrade operator` to upgrade the operator
+  components. Please note that this command will consider image overrides
+  defined in the local clusterctl config file.
+- If the admin doesn’t want to use clusterctl, they can use kubectl apply with
+  the latest version of the operator yaml that will be published in the
+  cluster-api release artifacts.
+- clusterctl will require a matching operator version. In the future, when
+  clusterctl move to beta/GA, we will reconsider supporting version skew
+  between clusterctl and the operator.
+
+##### Operator Delete
+
+- clusterctl will delete the operator as part of the `clusterctl delete --all`
+  command.
+- If the admin doesn’t want to use clusterctl, they can use kubectl delete.
+  However, it’s the admin’s responsibility to verify that there are no
+  providers running in the management cluster.
+
+#### Air gapped environment
+
+In order to use the operator in an air-gapped environment it is required
+
+1. Make the operator work in air-gapped environment
+   - To provide image overrides for the operator itself.
```

Comment:
```suggestion
   - To provide image overrides for the operator itself in order to pull the images
     from an accessible image repository.
```

Reply:
Are you envisioning using the current clusterctl image overrides for this?
Also, does this will setting will be local to the admin workstation?

Reply:
So in the first point of the [Operator Installation section](https://github.com/kubernetes-sigs/cluster-api/pull/3833/files#diff-a8b3fa322e4449932bb3e414c2582ca1acae4ae38e9fd9036e3df433f5b5528bR699-R702), we did mention that we are going to use the imageOverrides section for this.
 
But I do see your point that would cause another "local configuration" file that would need to be maintained separately. This feels like a chicken/egg problem here.
🤔 

So there are a couple of options that come to mind.
- The admin uses their own modified `operator.yaml` to install and upgrade the operator. This `operator.yaml` can be used within their own gitops practices this way. This can be viable but then things like `clusterctl upgrade plan` won't be able to inform if an upgrade is available or not.
- Currently we have a providers configuration in our `clusterctl.yaml` along with `imageOverrides` section, which we don't need anymore because of the declarative nature of our Provider API. However, since we still have other specific configs like the overrides layer, cert-manager-timeout and debug/log level there is still a need for `clusterctl.yaml` config. Maybe we can include an operator section. This section would be similar to the providers configuration in the sense that the admin can provide a URL to clusterctl from where to obtain the operator artifacts. This URL can point to a repository within the air-gapped environment and is under the control of the admins so that they can make the image overrides to that version directly. However, this feels a bit more round-about of the first option.

I'm leaning towards keeping it simple by using the imageOverrides defined in the local clusterctl config for now and documenting that users can maintain the `operator.yaml` in their own gitops process if they prefer another way.

Reply:
I'm ok with keeping it simple and use existing imageOverrides + Documentation.